### PR TITLE
Add argv and metadata support to C++ backend (Issue #2071)

### PR DESCRIPTION
**Pull request by steven-johnson**

### CODE REVIEW

File: src/CodeGen_C.cpp [hunk 1]
```diff
@@ -1489,6 +1489,252 @@ void CodeGen_C::forward_declare_type_if_needed(const Type &t) {
...
+        stream << "\nHALIDE_FUNCTION_ATTRS\nconst struct halide_filter_metadata_t *" << function_name << "_metadata();\n";
+        return;
+    }
+
+    auto map_name = [&metadata_name_map](const std::string &from) -> std::string {
+        auto it = metadata_name_map.find(from);
+        return it == metadata_name_map.end() ? from : it->second;
+    };
+
+    auto legalize_name = [](const std::string &from) -> std::string {
```
Comment (vksnk) [inline src/CodeGen_C.cpp:1539]: There is a `c_print_name` function, which does something similar. Is it possible to use it here instead?

**Review by vksnk — APPROVED** [review]
Thanks a lot for adding this!

File: src/CodeGen_C.cpp (hunk 1 above)
Comment (steven-johnson) [reply src/CodeGen_C.cpp]: Done

**Comment by steven-johnson** [comment]
Failures are unrelated.

**Pull request merged**